#! /bin/sh /usr/share/dpatch/dpatch-run
## 24_fix_pcm_no_control_status_mmap_on_32_compat_mode_64_bit_kernel.dpatch by  <crimsun@localhost.localdomain>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Fix PCM without control/status mmap on 32-bit compatible mode on 64-bit 
## DP: kernel

@DPATCH@

# HG changeset patch
# User crimsun@localhost.localdomain
# Node ID e515129cd3f19423d1beff5ed3dc1c174c49ef75
# Parent  adb0a046e5c35d98e5e09808348bb282546db22b
Subject: [PATCH] [UBUNTU alsa-lib] Fix PCM without control/status mmap on 32-bit compatible mode on 64-bit kernel

UpstreamStatus: Applied in upstream hg changeset 95447323ed91 [0]

This commit by Takashi Iwai fixes the update of appl_ptr via mmap_commit
when control/status structs are not mmapped, e.g., 32-bit compat mode on
64-bit kernel.

[0] http://hg-mirror.alsa-project.org/alsa-lib?cmd=changeset;node=95447323ed91dbadf64958ef4b364e9682aa60cb;style=gitweb

diff -r adb0a046e5c3 -r e515129cd3f1 src/pcm/pcm_hw.c
--- a/src/pcm/pcm_hw.c	Sun May 07 00:18:39 2006 -0400
+++ b/src/pcm/pcm_hw.c	Wed May 10 19:48:26 2006 -0400
@@ -924,6 +924,7 @@ static snd_pcm_sframes_t snd_pcm_hw_mmap
 		}
 	}
 	snd_pcm_mmap_appl_forward(pcm, size);
+	sync_ptr(hw, 0);
 #ifdef DEBUG_MMAP
 	fprintf(stderr, "appl_forward: hw_ptr = %li, appl_ptr = %li, size = %li\n", *pcm->hw.ptr, *pcm->appl.ptr, size);
 #endif
diff -r adb0a046e5c3 -r e515129cd3f1 src/pcm/pcm_plugin.c
--- a/src/pcm/pcm_plugin.c	Sun May 07 00:18:39 2006 -0400
+++ b/src/pcm/pcm_plugin.c	Wed May 10 19:48:26 2006 -0400
@@ -524,6 +524,9 @@ int snd_pcm_plugin_status(snd_pcm_t *pcm
 	snd_atomic_read_init(&ratom, &plugin->watom);
  _again:
 	snd_atomic_read_begin(&ratom);
+	/* sync with the latest hw and appl ptrs */
+	snd_pcm_plugin_avail_update(pcm);
+
 	err = snd_pcm_status(plugin->gen.slave, status);
 	if (err < 0) {
 		snd_atomic_read_ok(&ratom);
@@ -545,7 +548,7 @@ int snd_pcm_plugin_status(snd_pcm_t *pcm
 }
 
 snd_pcm_fast_ops_t snd_pcm_plugin_fast_ops = {
-	.status = snd_pcm_generic_status,
+	.status = snd_pcm_plugin_status,
 	.state = snd_pcm_generic_state,
 	.hwsync = snd_pcm_generic_hwsync,
 	.delay = snd_pcm_plugin_delay,
